### PR TITLE
Migration from Facebook and iCal

### PR DESCRIPTION
**Pull request by bordoni**
_Ref:_ [C#61940](https://central.tri.be/issues/61940)

### CODE REVIEW

File: src/Tribe/Aggregator/Migrate.php [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+		$aggregator = Tribe__Events__Aggregator::instance();
+
+		$html = '<p>' . esc_html__( 'Seems like you have old settings from before the Event Aggregator existed, the buttons bellow will allow you to migrate these legacy settings to Aggregator Records', 'the-events-calendar' );
+
+		if ( ! $this->is_facebook_migrated() && $this->has_facebook_setting() ) {
+			$html .= '<p style="display:inline-block;">' . get_submit_button( esc_html__( 'Migrate Facebook settings' ), 'secondary', 'tribe-migrate-facebook-settings', false ) . '<span class="spinner"></span></p>';
+		}
+
+		if ( ! $this->is_ical_migrated() && $this->has_ical_setting() ) {
+			$html .= '<p style="display:inline-block;">' . get_submit_button( esc_html__( 'Migrate iCal Settings' ), 'secondary', 'tribe-migrate-ical-settings', false ) . '<span class="spinner"></span></p>';
```
Comment (zbtirrell) [inline src/Tribe/Aggregator/Migrate.php]: needs translation domain

File: src/Tribe/Aggregator/Migrate.php [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+		) {
+			return false;
+		}
+
+		$aggregator = Tribe__Events__Aggregator::instance();
+
+		$html = '<p>' . esc_html__( 'Seems like you have old settings from before the Event Aggregator existed, the buttons bellow will allow you to migrate these legacy settings to Aggregator Records', 'the-events-calendar' );
+
+		if ( ! $this->is_facebook_migrated() && $this->has_facebook_setting() ) {
+			$html .= '<p style="display:inline-block;">' . get_submit_button( esc_html__( 'Migrate Facebook settings' ), 'secondary', 'tribe-migrate-facebook-settings', false ) . '<span class="spinner"></span></p>';
```
Comment (zbtirrell) [inline src/Tribe/Aggregator/Migrate.php]: needs translation domain

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | Comma required after last value in array declaration | [Migrate.php:159](https://github.com/moderntribe/the-events-calendar/blob/d52fe19ec1b6ed445035eec9e93fcf329a0a0c0a/src/Tribe/Aggregator/Migrate.php#L159)
:x: | Comma required after last value in array declaration | [Migrate.php:220](https://github.com/moderntribe/the-events-calendar/blob/d52fe19ec1b6ed445035eec9e93fcf329a0a0c0a/src/Tribe/Aggregator/Migrate.php#L220)
:warning: | Comment refers to a TODO task | [Migrate.php:301](https://github.com/moderntribe/the-events-calendar/blob/d52fe19ec1b6ed445035eec9e93fcf329a0a0c0a/src/Tribe/Aggregator/Migrate.php#L301)
:warning: | Comment refers to a TODO task | [Migrate.php:345](https://github.com/moderntribe/the-events-calendar/blob/d52fe19ec1b6ed445035eec9e93fcf329a0a0c0a/src/Tribe/Aggregator/Migrate.php#L345)
:warning: | Comment refers to a TODO task | [Migrate.php:379](https://github.com/moderntribe/the-events-calendar/blob/d52fe19ec1b6ed445035eec9e93fcf329a0a0c0a/src/Tribe/Aggregator/Migrate.php#L379)

> via jenkins/codesniffer-phpcs

File: src/Tribe/Aggregator/Record/List_Table.php [hunk 3]
```diff
@@ -483,6 +483,15 @@ public function column_source( $post ) {
 	}
 
 	public function column_imported( $post ) {
+		$record = Tribe__Events__Aggregator__Records::instance()->get_by_post_id( $post );
+		if ( 'scheduled' === $this->tab->get_slug() ) {
+			$has_child_record = $record->get_child_record_by_status( 'success', 1 );
+
+			if ( ! $has_child_record ) {
+				return $this->render( esc_html__( 'Never', 'the-events-calendar' ) );
```
Comment (zbtirrell) [inline src/Tribe/Aggregator/Record/List_Table.php]: @leahkoerper should this be "On Demand"?

File: src/Tribe/Aggregator.php [hunk 4]
```diff
@@ -94,6 +94,7 @@ private function __construct() {
 
 		// Loads the Required Classes and saves them as proprieties
 		$this->meta_box        = Tribe__Events__Aggregator__Meta_Box::instance();
+		$this->migrate        = Tribe__Events__Aggregator__Migrate::instance();
```
Comment (zbtirrell) [inline src/Tribe/Aggregator.php]: spacing seems off here

File: src/Tribe/Aggregator/Migrate.php [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+	 *
+	 * @param  string $frequency iCal Frequency
+	 * @return string            EA Frequency
+	 */
+	private function convert_ical_frequency( $frequency ) {
+		$results = Tribe__Events__Aggregator__Cron::instance()->get_frequency( array( 'id' => $frequency ) );
+
+		// Return to the Default
+		if ( empty( $results ) ) {
+			return 'every30mins';
```
Comment (zbtirrell) [inline src/Tribe/Aggregator/Migrate.php]: this is the iCal Importer default? The new EA default is daily, should we slide those that way, or is this really the appropriate option?

Comment (zbtirrell) [inline src/Tribe/Aggregator/Migrate.php]: or is this just when there is no match?

File: src/resources/js/aggregator-admin-legacy-settings.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
+var tribe_aggregator = tribe_aggregator || {};
```
Comment (zbtirrell) [inline src/resources/js/aggregator-admin-legacy-settings.js]: surprised this JS isn't minified

Comment (bordoni) [inline src/resources/js/aggregator-admin-legacy-settings.js]: Good catch.

File: src/Tribe/Aggregator/Migrate.php (hunk 5 above)
Comment (bordoni) [inline src/Tribe/Aggregator/Migrate.php]: When there is no match means, FB 30mins and 15mins, nothing else.

File: src/Tribe/Aggregator/Migrate.php [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+				'file'         => null,
+				'keywords'     => null,
+				'location'     => null,
+				'start'        => null,
+				'radius'       => null,
+				'source'       => $source,
+				'content_type' => null,
+				'is_legacy'    => true,
+				'import_id'    => null,
+				'post_status'  => $settings->post_status,
```
Comment (borkweb) [inline src/Tribe/Aggregator/Migrate.php]: Legacy iCal imports can have keywords, location, start, and radius. We should make sure those values are added here if they were set.

File: src/Tribe/Aggregator/Migrate.php [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+				$args['post_status'] = Tribe__Events__Aggregator__Records::$status->schedule;
+				wp_update_post( $args );
+			}
+		}
+
+		/**
+		 * @todo Create a real Logic for Messaging what happened
+		 */
+		$response->status = true;
+		$response->text = esc_html__( 'Succesfully imported the Facebook Settings into Aggregator Records.', 'the-events-calendar' );
```
Comment (borkweb) [inline src/Tribe/Aggregator/Migrate.php]: This should say iCal rather than Facebook

File: src/Tribe/Aggregator/Migrate.php [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+		$response = (object) array(
+			'status' => false,
+			'text' => esc_html__( 'Error, a unknown bug happened and it was impossible to migrate the old iCal Settings to Event Aggregator, try again later.', 'the-events-calendar' ),
+		);
+
+		$post_type = get_post_type_object( Tribe__Events__Main::POSTTYPE );
+
+		if ( empty( $post_type->cap->edit_posts ) || ! current_user_can( $post_type->cap->edit_posts ) ) {
+			$response->status = false;
+			$response->text = esc_html__( 'You do not have permission to migrate the iCal Facebook Settings to Event Aggregator', 'the-events-calendar' );
```
Comment (borkweb) [inline src/Tribe/Aggregator/Migrate.php]: Remove the word "Facebook"

File: src/Tribe/Aggregator/Migrate.php [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+		foreach ( $settings->ids as $source ) {
+			$record = Tribe__Events__Aggregator__Records::instance()->get_by_origin( $origin );
+
+			/**
+			 * @todo Include the Deactivated logic
+			 */
+			$meta = array(
+				'origin'       => $origin,
+				'type'         => 'schedule',
+				'frequency'    => $this->convert_ical_frequency( $settings->frequency ),
```
Comment (borkweb) [inline src/Tribe/Aggregator/Migrate.php]: Should `convert_ical_frequency` be used here?

File: src/Tribe/Aggregator/Migrate.php (hunk 5 above)
Comment (borkweb) [inline src/Tribe/Aggregator/Migrate.php]: My vote is for `daily` when there's no match.

Comment (bordoni) [inline src/Tribe/Aggregator/Migrate.php]: It makes no sense to convert a 15min and 30 minutes frequency to daily,  I'm going towards the closest.

Comment (borkweb) [inline src/Tribe/Aggregator/Migrate.php]: Ahhhhh...I totally misread your comment above mine. Even after re-reading it multiple times, it wasn't clear to me. I get it now.

**Comment by borkweb** [comment]
:partyparrot:

**Pull request merged**